nodefs/bridge.go

```diff
@@ -0,0 +1,496 @@
+		RawFileSystem: fuse.NewDefaultRawFileSystem(),
+	}
+
+	if opts != nil {
+		bridge.options = *opts
+	} else {
+		oneSec := time.Second
+		bridge.options.EntryTimeout = &oneSec
+		bridge.options.AttrTimeout = &oneSec
+	}
+
+	bridge.root = &Inode{
+		nodeID:      1,
+		lookupCount: 1,
+		mode:        fuse.S_IFDIR,
+		children:    make(map[string]*Inode),
+		parents:     nil,
+		node:        root,
+		bridge:      bridge,
+	}
+	root.setInode(bridge.root)
+	bridge.nodes = append(bridge.nodes,
+		mapEntry{},
+		// ID 1 is always the root.
+		mapEntry{inode: bridge.root})
+
+	// Fh 0 means no file handle.
+	bridge.files = []fileEntry{{}}
+	return bridge
+}
+
+func (b *rawBridge) inode(id uint64, fh uint64) (*Inode, fileEntry) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.nodes[id].inode, b.files[fh]
+}
+
+func (b *rawBridge) Lookup(header *fuse.InHeader, name string, out *fuse.EntryOut) (status fuse.Status) {
+	parent, _ := b.inode(header.NodeId, 0)
+
+	child, code := parent.node.Lookup(context.TODO(), name, out)
+	if !code.Ok() {
+		if b.options.NegativeTimeout != nil {
+			out.SetEntryTimeout(*b.options.NegativeTimeout)
+		}
+		return code
+	}
+
+	b.addNewChild(parent, name, child, out)
+	b.setEntryOutTimeout(out)
+	return fuse.OK
+}
+
+func (b *rawBridge) Rmdir(header *fuse.InHeader, name string) fuse.Status {
+	parent, _ := b.inode(header.NodeId, 0)
+	code := parent.node.Rmdir(context.TODO(), name)
+	if code.Ok() {
+		parent.RmChild(name)
+	}
+	return code
+
+}
+
+func (b *rawBridge) Unlink(header *fuse.InHeader, name string) fuse.Status {
+	parent, _ := b.inode(header.NodeId, 0)
+	code := parent.node.Unlink(context.TODO(), name)
+	if code.Ok() {
+		parent.RmChild(name)
+	}
+	return code
+}
+
+func (b *rawBridge) Mkdir(input *fuse.MkdirIn, name string, out *fuse.EntryOut) (code fuse.Status) {
+	parent, _ := b.inode(input.NodeId, 0)
+
+	child, code := parent.node.Mkdir(context.TODO(), name, input.Mode, out)
+	if !code.Ok() {
+		return code
+	}
+
+	if out.Attr.Mode&^07777 != fuse.S_IFDIR {
+		log.Panicf("Mkdir: mode must be S_IFDIR (%o), got %o", fuse.S_IFDIR, out.Attr.Mode)
+	}
+
+	b.addNewChild(parent, name, child, out)
+	b.setEntryOutTimeout(out)
+	return fuse.OK
+}
+
+func (b *rawBridge) Mknod(input *fuse.MknodIn, name string, out *fuse.EntryOut) (code fuse.Status) {
+	parent, _ := b.inode(input.NodeId, 0)
+
+	child, code := parent.node.Mknod(context.TODO(), name, input.Mode, input.Rdev, out)
+	if !code.Ok() {
+		return code
+	}
+
+	b.addNewChild(parent, name, child, out)
+	b.setEntryOutTimeout(out)
+	return fuse.OK
+}
+
+func (b *rawBridge) addNewChild(parent *Inode, name string, child *Inode, out *fuse.EntryOut) {
+	lockNodes(parent, child)
+	parent.setEntry(name, child)
+	b.mu.Lock()
+	if child.nodeID == 0 {
+		b.registerInode(child)
+	}
+	out.NodeId = child.nodeID
+	// NOSUBMIT - or should let FS expose Attr.Ino? This makes
+	// testing semantics hard though, because os.Lstat doesn't
+	// reflect the FUSE FS
+	out.Attr.Ino = child.nodeID
```

Comment:
Yes please let the FS decide about the the inode number. How about

```
if out.Attr.Ino == 0 {
out.Attr.Ino = \<something\>
}
```

Also I don't think the nodeID is suitable, because it gets recycled quickly. Imagine `rsync -aH` looking for hard links. Because the nodeID is quickly recycled, it will find lots of hard linked files, and the resulting copy will be corrupted..

Reply:
Maybe
```
out.Attr.Ino = atomic.AddUint64(b.inum, 1)
```
?

(whatever we do needs to be replicated in GetAttr)

Reply:
if we generate a new inode number for each call to NewInode(), we're still "wrong". You're saying it is better to pretend files that are really hardlinked are different?

We could generate the Inode number from the FileID, but the problem is that inode must be 64 bit, while FileID is 96 bits (64 for inode, 32 for device.)

Reply:
alternatively, we could do: 

- let FileID be a u64
- use this u64 as NodeID to communicate with the kernel. 

This requires an extra translation layer from (dev,ino) <-> nodeID in the loopback FS,

Reply:
> You're saying it is better to pretend files that are really hardlinked are different?

Yes, that's extra disk space used, the other one is data loss.

Reply:
About making FileID = nodeID = inum (inode number in stat): Actually, sounds good. The additional (dev,ino) <-> nodeID map in loopbackfs pays for itself because bridge can drop the the fileID <-> nodeID map `byFileID`.

OTOH, the kernel and libfuse people thought it is useful that nodeID and inum can have different values. Why? Maybe it's just so that libfuse can put a raw memory pointer into nodeID, without exposing it in the stat() inum, which would be ugly for several reasons.

We don't want to use raw memory pointers as nodeID, so I guess we can drop the disctinction between nodeID and inum, and simplify the whole thing a little.

If an FS does not want to use hard links, or cannot provide stable inums, it should return zero, and bridge can do `Ino = atomic.AddUint64(b.inum, 1)`. But I wonder if the kernel gets unhappy if it always gets a new nodeID for each LOOKUP. Will test.

Reply:
there is an additional factor, which is the cost of a map[uint64] vs lookup in []inodeEntry

it's probablly fast enough to not be relevant, though.

Reply:
for loopbackFS, maybe we can do something like

  (dev << 32 | inode64)

this would work well for filesystems that hand inode numbers sequentially. 

(would not work well for stacking go-fuse FSes)

Reply:
That's true, the nodeID lookup is super-fast at the moment due to the tight packing.

For the inode64 mapping, my thought for gocryptfs was to check the device number of the root dir, and pass all inode numbers with this device number unchanged. All other device numbers could get the (dev << 32 | inode64) treatment.

Once stacking starts, this needs a proper map, but that's not too expensive: 96-bit key -> 64-bit value = 20 bytes per inode number?

Reply:
I used to have raw pointers in the NodeID field, and pass them on as Ino in Stat. There were 2 problems:

- the race detector gets really confused
- older 32-bit software (notably some JDKs) would barf with ERANGE when trying to stuff a 64 bit ino into a 32 bit stat_t.

Reply:
EncFS also used to use raw pointers [till recently](https://github.com/vgough/encfs/commit/e2f0f8e3c68604a46d8665c3b0125bb35a0d9181). It also had a subtle bug with concurrent file open, which meant a node that still was in use by the kernel could get deallocated. Was fun to debug because it caused random segfaults: https://github.com/vgough/encfs/issues/214 . 0 of 5 stars for raw pointers from me.